// neovim/lib/api/Base.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const logger_1 = require("../utils/logger");
// Instead of dealing with multiple inheritance (or lackof), just extend EE
// Only the Neovim API class should use EE though
class BaseApi extends events_1.EventEmitter {
    constructor({ transport, data, logger, metadata }) {
        super();
        this.transport = transport;
        this.data = data;
        this.logger = logger || logger_1.logger;
        if (metadata) {
            Object.defineProperty(this, 'metadata', { value: metadata });
        }
    }
    equals(other) {
        try {
            return this.data.toString() === other.data.toString();
        }
        catch (e) {
            return false;
        }
    }
    request(name, args = []) {
        return __awaiter(this, void 0, void 0, function* () {
            // `this._isReady` is undefined in ExtType classes (i.e. Buffer, Window, Tabpage)
            // But this is just for Neovim API, since it's possible to call this method from Neovim class
            // before transport is ready.
            // Not possible for ExtType classes since they are only created after transport is ready
            yield this._isReady;
            this.logger.debug(`request -> neovim.api.${name}`);
            return new Promise((resolve, reject) => {
                this.transport.request(name, args, (err, res) => {
                    this.logger.debug(`response -> neovim.api.${name}: ${res}`);
                    if (err) {
                        reject(new Error(`${name}: ${err[1]}`));
                    }
                    else {
                        resolve(res);
                    }
                });
            });
        });
    }
    _getArgsByPrefix(...args) {
        const _args = [];
        // Check if class is Neovim and if so, should not send `this` as first arg
        if (this.prefix !== 'nvim_') {
            _args.push(this);
        }
        return _args.concat(args);
    }
    /** Retrieves a scoped variable depending on type (using `this.prefix`) */
    getVar(name) {
        const args = this._getArgsByPrefix(name);
        return this.request(`${this.prefix}get_var`, args).then(res => res, err => {
            if (err && err.message && err.message.includes('not found')) {
                return null;
            }
            throw err;
        });
    }
    /** Set a scoped variable */
    setVar(name, value) {
        const args = this._getArgsByPrefix(name, value);
        return this.request(`${this.prefix}set_var`, args);
    }
    /** Delete a scoped variable */
    deleteVar(name) {
        const args = this._getArgsByPrefix(name);
        return this.request(`${this.prefix}del_var`, args);
    }
    /** Retrieves a scoped option depending on type of `this` */
    getOption(name) {
        const args = this._getArgsByPrefix(name);
        return this.request(`${this.prefix}get_option`, args);
    }
    /** Set scoped option */
    setOption(name, value) {
        const args = this._getArgsByPrefix(name, value);
        return this.request(`${this.prefix}set_option`, args);
    }
    // TODO: Is this necessary?
    /** `request` is basically the same except you can choose to wait forpromise to be resolved */
    notify(name, args) {
        this.logger.debug(`notify -> neovim.api.${name}`);
        this.transport.notify(name, args);
    }
}
exports.BaseApi = BaseApi;
